src/hotspot/share/classfile/loaderConstraints.cpp

```diff
@@ -27,80 +27,158 @@
 #include "classfile/classLoaderDataGraph.hpp"
 #include "classfile/dictionary.hpp"
 #include "classfile/loaderConstraints.hpp"
+#include "classfile/placeholders.hpp"
 #include "logging/log.hpp"
 #include "memory/resourceArea.hpp"
 #include "oops/klass.inline.hpp"
 #include "oops/oop.inline.hpp"
 #include "runtime/handles.inline.hpp"
+#include "runtime/mutexLocker.hpp"
 #include "runtime/safepoint.hpp"
-#include "utilities/hashtable.inline.hpp"
+#include "utilities/resourceHash.hpp"
+
+// Implementation Classes for Loader Constraints
+
+class LoaderConstraint : public CHeapObj<mtClass> {
+  InstanceKlass*         _klass;
+  // Loader constraints enforce correct linking behavior.
+  // Thus, it really operates on ClassLoaderData which represents linking domain,
+  // not class loaders.
+  GrowableArray<ClassLoaderData*>*  _loaders;                // initiating loaders
+ public:
+  LoaderConstraint(InstanceKlass* klass, oop class_loader1, oop class_loader2) :
+     _klass(klass) {
+    _loaders = new (ResourceObj::C_HEAP, mtClass) GrowableArray<ClassLoaderData*>(10, mtClass);
+    set_loader(class_loader1);
+    set_loader(class_loader2);
+  }
 
-void LoaderConstraintEntry::set_loader(int i, oop p) {
-  set_loader_data(i, ClassLoaderData::class_loader_data(p));
-}
+  LoaderConstraint() { delete _loaders; }
```

Comment:
A typo (should be `~LoaderConstraint` ??)

Reply:
Yes, that's a bad typo.